### PR TITLE
Feature/al 44 raw send and receive

### PR DESCRIPTION
**Pull request by stesmi**

### CODE REVIEW

File: src/Zforce.cpp [hunk 1]
```diff
@@ -123,6 +124,110 @@ int Zforce::Write(uint8_t* payload)
...
+      this->remainingRawLength = fullAsn1MessageLength - i2cPayloadLength;
+      *remainingLength = this->remainingRawLength;
+      *receivedLength = i2cPayloadLength;
+      return &buffer[2]; // Skipping the i2c header in the response.
+    }
+    else
+    {
+      // Since this is the second or later invocation, we do NOT parse the ASN.1, since it
+      // may well be in the middle of some data, and we already know the lengths.
+      this->remainingRawLength -= asn1PayloadLength;
```
Comment (twarenius) [inline src/Zforce.cpp]: Where is `asn1PayloadLength` coming from? It doesn't compile without it.

**Review by twarenius — CHANGES_REQUESTED** [review]

File: src/Zforce.cpp (hunk 1 above)
Comment (stesmi) [reply src/Zforce.cpp]: Had renamed it, and forgot to change that.

File: src/Zforce.cpp [hunk 2]
```diff
@@ -123,6 +124,110 @@ int Zforce::Write(uint8_t* payload)
...
+        }
+        else if (numberOfLengthBytes != 1)
+        {
+          // The data is most likely corrupted. Valid numbers are 1 and 2.
+          *receivedLength = 0;
+          *remainingLength = 0;
+          return nullptr;
+        }
+      }
+      uint16_t fullAsn1MessageLength = asn1AfterHeaderLength + asn1HeaderLength
```
Comment (twarenius) [inline src/Zforce.cpp]: It still doesn't compile. One might think the Arduino GCC compiler is picky but it is missing a semicolon.

**Review by twarenius — CHANGES_REQUESTED** [review]

**Review by twarenius — DISMISSED** [review]
Looks good. I have test run both the SendRaw and ReceiveRaw methods and they both seem to work fine though I haven't been able to produce a scenario for testing the ReceiveRaw method with a large enough ASN.1 response to require multiple invocations.

**Comment by stesmi** [comment]
> 

I don't like it, either. That means we need to check the code even better than if we could test it.

**Review by braide — APPROVED** [review]

**Review by twarenius — APPROVED** [review]
I have test run the latest changes now and it works as expected. Worth noting though that the case where the ASN.1 payload exceeds the I2C payload size has not been tested.

**Pull request merged**